### PR TITLE
Add a maximum size for well-known lookups.

### PR DESCRIPTION
**Pull request by clokep**
I noticed we already had a way to enforce a maximum size when reading a response from Twisted. There's no reason for a .well-known lookup to be large so we should bail after a short amount of data. I chose 50kb in here, but that's somewhat arbitrary, the maximum reasonable size for a well-known response is:

```python
len("""{
    "m.server": "<maximum domain length>:65535"
}
""")
```

Note that the maximum domain length is 253 characters. So this leads to roughly 28 + 253 < 300 bytes. We can bikeshed the maximum size if you'd like.

Note that I don't *think* there's anything else useful to put in there.

### CODE REVIEW

**Comment by tulir** [comment]
> Note that I don't think there's anything else useful to put in there.

Some people might handle all `/.well-known/matrix/*` paths as one and return the same JSON object regardless of the file name. 50kb limit should still be fine though

File: synapse/http/client.py [hunk 1]
```diff
@@ -748,7 +751,11 @@ def _timeout_to_request_timed_out_error(f: Failure):
     return f
 
 
-class _ReadBodyToFileProtocol(protocol.Protocol):
+class BodyExceededMaxSize(Exception):
```
Comment (richvdh) [inline synapse/http/client.py:754]: we don't seem to actually raise this?

File: synapse/http/client.py [hunk 2]
```diff
@@ -782,7 +789,7 @@ def connectionLost(self, reason: Failure) -> None:
             self.deferred.errback(reason)
 
 
-def readBodyToFile(
+def readBodyWithMaxSize(
```
Comment (richvdh) [inline synapse/http/client.py:792]: is there a reason for the unpythonic camelCase name here? we should probably fix that while we're here.

Comment (richvdh) [inline synapse/http/client.py:792]: also the docstring should document what happens if the file is too big

File: synapse/http/federation/well_known_resolver.py [hunk 3]
```diff
@@ -53,6 +55,9 @@
 # lower bound for .well-known cache period
 WELL_KNOWN_MIN_CACHE_PERIOD = 5 * 60
 
+# The maximum size (in bytes) to allow a well-known file to be.
+WELL_KNOW_MAX_SIZE = 50 * 1024  # 50 KiB
```
Comment (richvdh) [inline synapse/http/federation/well_known_resolver.py:59]: `well_knowN`

**Review by richvdh — CHANGES_REQUESTED** [review]

File: synapse/http/client.py (hunk 1 above)
Comment (clokep) [reply synapse/http/client.py:754]: Arg. Good catch. 🤦

File: synapse/http/client.py (hunk 2 above)
Comment (clokep) [reply synapse/http/client.py:792]: I think it is to match the method from Twisted call `readBody`. I can update it though!

File: synapse/http/client.py [hunk 4]
```diff
@@ -782,12 +783,15 @@ def connectionLost(self, reason: Failure) -> None:
             self.deferred.errback(reason)
 
 
-def readBodyToFile(
+def read_body_with_max_size(
```
Comment (clokep) [inline synapse/http/client.py:786]: I wanted to make this async and handle the `Deferred` logic internally, but the call from the `MatrixFederationHttpClient` adds a timeout, which involves having both a timeout and a reactor. The timeout could be made optional easily, but needed both or neither is kind of meh.

**Review by richvdh — APPROVED** [review]
lgtm

**Review by ddsmatt — COMMENTED** [review]
Reviewed changes

**Pull request merged**